# DATA QUALITY ASSESSMENT

We have been asked to evaluate the data quality of Transactions, Customer Demographics and Customer Address

In [81]:
import pandas as pd

# Transactions

In [82]:
transactions = pd.read_excel(r"KPMG.xlsx", sheet_name='Transactions', header=1)
print(transactions.head())

   transaction_id  product_id  customer_id transaction_date  online_order   
0               1           2         2950       2017-02-25           0.0  \
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size   
0     Approved           Solex     Standard        medium       medium  \
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62        

In [83]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [84]:
transactions.shape

(20000, 13)

In [85]:
transactions['product_first_sold_date'].head()

0    41245.0
1    41701.0
2    36361.0
3    36145.0
4    42226.0
Name: product_first_sold_date, dtype: float64

In [86]:
transactions.duplicated().sum() #to see if there are any duplicated rows

0

In [87]:
transactions.isnull().sum()  #check for missing values

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [88]:
#Number of rows that contain no NaN values in all its columns
null_rows = transactions.notna().all(axis=1).sum()
print('No. of records with non NaN:', null_rows)
print('% of Missing data:', round((20000 - null_rows)*100 / 20000,1))

No. of records with non NaN: 19445
% of Missing data: 2.8


In [89]:
transactions.nunique() #to see the unique values in each column

transaction_id             20000
product_id                   101
customer_id                 3494
transaction_date             364
online_order                   2
order_status                   2
brand                          6
product_line                   4
product_class                  3
product_size                   3
list_price                   296
standard_cost                103
product_first_sold_date      100
dtype: int64

In [90]:
transactions.describe().T

count                        mean   
transaction_id           20000.0                     10000.5  \
product_id               20000.0                    45.36465   
customer_id              20000.0                  1738.24605   
transaction_date           20000  2017-07-01 14:08:05.280000   
online_order             19640.0                    0.500458   
list_price               20000.0                 1107.829449   
standard_cost            19803.0                  556.046951   
product_first_sold_date  19803.0                38199.776549   

                                         min                  25%   
transaction_id                           1.0              5000.75  \
product_id                               0.0                 18.0   
customer_id                              1.0               857.75   
transaction_date         2017-01-01 00:00:00  2017-04-01 00:00:00   
online_order                             0.0                  0.0   
list_price                             12.01               575.27   
standard_cost                           7.21               215.14   
product_first_sold_date              33259.0              35667.0   

                                         50%                  75%   
transaction_id                       10000.5             15000.25  \
product_id                              44.0                 72.0   
customer_id                           1736.0               2613.0   
transaction_date         2017-07-03 00:00:00  2017-10-02 00:00:00   
online_order                             1.0                  1.0   
list_price                           1163.89               1635.3   
standard_cost                         507.58                795.1   
product_first_sold_date              38216.0              40672.0   

                                         max          std  
transaction_id                       20000.0  5773.647028  
product_id                             100.0     30.75359  
customer_id                           5034.0  1011.951046  
transaction_date         2017-12-30 00:00:00          NaN  
online_order                             1.0     0.500013  
list_price                           2091.47   582.825242  
standard_cost                        1759.85    405.95566  
product_first_sold_date              42710.0   2875.20111

In [91]:
transactions['order_status'].value_counts()

order_status
Approved     19821
Cancelled      179
Name: count, dtype: int64

In [92]:
transactions['product_line'].value_counts()

product_line
Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: count, dtype: int64

In [93]:
transactions['brand'].value_counts()

brand
Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: count, dtype: int64

### Summary
* 20,000 rows, 3494 unique customer_ids
* We can see that there are some missing values in the columns : 'online_order', 'brand', 'product_line ', 'product_class', 'product_size', 'standard_cost', 'product_first_sold_date'. In total, the missing values represent 2.8% of all the rows so it may be appropiate to drop them
* We can also notice that the 'product_first_sold_date' column has a float datatype although it's name describes a date format. However, after reviewing the column it does not represent the product first sold date. It appears to have redundant values, therefore we should drop the column.
* There are not duplicated rows
* Transaction_id, product_id and customer_id should be strings as they represent ids

# CUSTOMER DEMOGRAPHIC

In [94]:
cus_demographic = pd.read_excel(r"KPMG.xlsx", sheet_name='CustomerDemographic', header=1)
cus_demographic.head()

customer_id      first_name  last_name  gender   
0            1         Laraine  Medendorp       F  \
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases                  DOB   
0                                   93  1953-10-12 00:00:00  \
1                                   81  1980-12-16 00:00:00   
2                                   61  1954-01-20 00:00:00   
3                                   33  1961-10-03 00:00:00   
4                                   56  1977-05-13 00:00:00   

                job_title job_industry_category     wealth_segment   
0     Executive Secretary                Health      Mass Customer  \
1  Administrative Officer    Financial Services      Mass Customer   
2      Recruiting Manager              Property      Mass Customer   
3                     NaN                    IT      Mass Customer   
4           Senior Editor                   NaN  Affluent Customer   

  deceased_indicator                                            default   
0                  N                                                 "'  \
1                  N                       <script>alert('hi')</script>   
2                  N                                2018-02-01 00:00:00   
3                  N  () { _; } >_[$($())] { touch /tmp/blns.shellsh...   
4                  N                                                NIL   

  owns_car  tenure  
0      Yes    11.0  
1      Yes    16.0  
2      Yes    15.0  
3       No     7.0  
4      Yes     8.0

In [95]:
cus_demographic.shape

(4000, 13)

In [96]:
cus_demographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          4000 non-null   int64  
 1   first_name                           4000 non-null   object 
 2   last_name                            3875 non-null   object 
 3   gender                               4000 non-null   object 
 4   past_3_years_bike_related_purchases  4000 non-null   int64  
 5   DOB                                  3913 non-null   object 
 6   job_title                            3494 non-null   object 
 7   job_industry_category                3344 non-null   object 
 8   wealth_segment                       4000 non-null   object 
 9   deceased_indicator                   4000 non-null   object 
 10  default                              3698 non-null   object 
 11  owns_car                      

In [97]:
cus_demographic.isnull().sum() #to see the missing values in column

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

In [98]:
#Number of rows that contain no NaN values in all its columns
null_rows_demo = cus_demographic.notna().all(axis=1).sum()
print('No. of records with non NaN:', null_rows_demo)
print('% of Missing data:', round((4000 - null_rows_demo)*100 / 4000,1))

No. of records with non NaN: 2630
% of Missing data: 34.2


In [99]:
cus_demographic.duplicated().sum() #to see if there are any duplicated items

0

In [100]:
cus_demographic['DOB'] = pd.to_datetime(cus_demographic['DOB'])

In [101]:
cus_demographic.nunique()

customer_id                            4000
first_name                             3139
last_name                              3725
gender                                    6
past_3_years_bike_related_purchases     100
DOB                                    3448
job_title                               195
job_industry_category                     9
wealth_segment                            3
deceased_indicator                        2
default                                  90
owns_car                                  2
tenure                                   22
dtype: int64

In [102]:
cus_demographic.describe().T

count                        mean   
customer_id                          4000.0                      2000.5  \
past_3_years_bike_related_purchases  4000.0                       48.89   
DOB                                    3913  1977-07-12 23:56:41.277792   
tenure                               3913.0                   10.657041   

                                                     min                  25%   
customer_id                                          1.0              1000.75  \
past_3_years_bike_related_purchases                  0.0                 24.0   
DOB                                  1843-12-21 00:00:00  1968-01-22 00:00:00   
tenure                                               1.0                  6.0   

                                                     50%                  75%   
customer_id                                       2000.5              3000.25  \
past_3_years_bike_related_purchases                 48.0                 73.0   
DOB                                  1977-07-22 00:00:00  1987-02-28 00:00:00   
tenure                                              11.0                 15.0   

                                                     max          std  
customer_id                                       4000.0  1154.844867  
past_3_years_bike_related_purchases                 99.0    28.715005  
DOB                                  2002-03-11 00:00:00          NaN  
tenure                                              22.0     5.660146

In [103]:
cus_demographic['gender'].value_counts()

gender
Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: count, dtype: int64

In [104]:
cus_demographic['default'].value_counts()

default
100                                       113
1                                         112
-1                                        111
-100                                       99
١٢٣                                     53
                                         ... 
testâ testâ«                               31
/dev/null; touch /tmp/blns.fail ; echo     30
âªâªtest⪠                                29
ì¸ëë°í ë¥´                                 27
,ãã»:*:ã»ãâ( â» Ï â» )ãã»:*:ã»ãâ           25
Name: count, Length: 90, dtype: int64

In [124]:
cus_demographic['DOB'].min()

Timestamp('1843-12-21 00:00:00')

In [126]:
cus_demographic['DOB'].value_counts()

DOB
1978-01-30    7
1964-07-08    4
1962-12-17    4
1978-08-19    4
1977-05-13    4
             ..
1989-06-16    1
1998-09-30    1
1985-03-11    1
1989-10-23    1
1991-11-05    1
Name: count, Length: 3448, dtype: int64

In [106]:
cus_demographic['job_title'].value_counts()

job_title
Business Systems Development Analyst    45
Tax Accountant                          44
Social Worker                           44
Internal Auditor                        42
Recruiting Manager                      41
                                        ..
Database Administrator I                 4
Health Coach I                           3
Health Coach III                         3
Research Assistant III                   3
Developer I                              1
Name: count, Length: 195, dtype: int64

In [107]:
cus_demographic['job_industry_category'].value_counts()

job_industry_category
Manufacturing         799
Financial Services    774
Health                602
Retail                358
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: count, dtype: int64

In [108]:
cus_demographic['wealth_segment'].value_counts()

wealth_segment
Mass Customer        2000
High Net Worth       1021
Affluent Customer     979
Name: count, dtype: int64

In [109]:
cus_demographic['deceased_indicator'].value_counts()

deceased_indicator
N    3998
Y       2
Name: count, dtype: int64

In [110]:
cus_demographic['owns_car'].value_counts()

owns_car
Yes    2024
No     1976
Name: count, dtype: int64

In [111]:
cus_demographic['tenure'].value_counts()

tenure
7.0     235
5.0     228
11.0    221
10.0    218
16.0    215
8.0     211
18.0    208
12.0    202
9.0     200
14.0    200
6.0     192
13.0    191
4.0     191
17.0    182
15.0    179
1.0     166
3.0     160
19.0    159
2.0     150
20.0     96
22.0     55
21.0     54
Name: count, dtype: int64

### Summary
* 4000 rows & 4000 unique customer_ids
* Missing values in the columns : 'last_name', 'DOB', 'job_title', 'job_industry_category', 'default', 'ternure'. In total, the missing values represent 34.2%, so it might be appropiate to replace those values
* 'DOB' has an inconsistent value: 1843 year of birth
* 'DOB' datatype is object when it should be in date format
* There are no duplicated values
* The default colum have redundant values. It seems to serve no purpose so it may be appropiate to be dropped from the table.
* The 'gender' column has double entries: 'Male', 'Female', 'M', 'F'. & 'U' should be 'Unidentified'.
* Customer_id is an integer when it should be a string

# CUSTOMER ADDRESS 

In [112]:
cust_address = pd.read_excel(r"KPMG.xlsx", sheet_name='CustomerAddress', header=1)
cust_address.head()

customer_id              address  postcode            state    country   
0            1   060 Morning Avenue      2016  New South Wales  Australia  \
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [113]:
cust_address.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [114]:
cust_address.shape

(3999, 6)

In [115]:
cust_address.isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

In [116]:
cust_address.duplicated().sum()

0

In [117]:
cust_address.nunique()

customer_id           3999
address               3996
postcode               873
state                    5
country                  1
property_valuation      12
dtype: int64

In [118]:
cust_address.describe().T

count         mean          std     min     25%     50%   
customer_id         3999.0  2003.987997  1154.576912     1.0  1004.5  2004.0  \
postcode            3999.0  2985.755939   844.878364  2000.0  2200.0  2768.0   
property_valuation  3999.0     7.514379     2.824663     1.0     6.0     8.0   

                       75%     max  
customer_id         3003.5  4003.0  
postcode            3750.0  4883.0  
property_valuation    10.0    12.0

In [119]:
cust_address['address'].value_counts()

address
3 Mariners Cove Terrace      2
3 Talisman Place             2
64 Macpherson Junction       2
359 Briar Crest Road         1
4543 Service Terrace         1
                            ..
5063 Shopko Pass             1
09 Hagan Pass                1
87897 Lighthouse Bay Pass    1
294 Lawn Junction            1
320 Acker Drive              1
Name: count, Length: 3996, dtype: int64

In [120]:
cust_address['postcode'].value_counts()

postcode
2170    31
2155    30
2145    30
2153    29
3977    26
        ..
3808     1
3114     1
4721     1
4799     1
3089     1
Name: count, Length: 873, dtype: int64

In [121]:
cust_address['state'].value_counts()

state
NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: count, dtype: int64

In [122]:
cust_address['country'].value_counts()

country
Australia    3999
Name: count, dtype: int64

In [123]:
cust_address['property_valuation'].value_counts()

property_valuation
9     647
8     646
10    577
7     493
11    281
6     238
5     225
4     214
12    195
3     186
1     154
2     143
Name: count, dtype: int64

### Summary
* 3999 rows & 3999 unique customer_ids
* No missing values & No duplicated rows
* The Customer Address Table is consisent and accurate. The table seems to contain all the customer address information about customers in Australia
* State Values (VIC & NSW) are duplicated (Victoria & New South Wales)
* Customer_id is an integer when it should be a string